action.js

```diff
@@ -23,27 +23,33 @@ import { memoize } from "./memo.js";
  *      single, object argument into a list of sorted values
  */
 const makeAction = (type, payloadKeys) => {
-  payloadKeys.sort();
-  const memoized = memoize((...values) => {
-    const payload = {};
-    values.forEach((v, i) => {
-      payload[payloadKeys[i]] = v;
+  if (payloadKeys) {
+    payloadKeys.sort();
+    const memoized = memoize((...values) => {
+      const payload = {};
+      values.forEach((v, i) => {
+        payload[payloadKeys[i]] = v;
+      });
+      return { type, payload };
     });
-    return { type, payload };
-  });
-  return (obj) => memoized(...payloadKeys.map((k) => obj[k]));
+    return (obj) => memoized(...payloadKeys.map((k) => obj[k]));
+  } else {
+    return (ary) => {
+      return { type, payload: ary };
+    };
+  }
 };
 
 const noop = Object.freeze({ type: "noop" });
 
 export default {
   noop,
-  initialize: makeAction("initialize", ["players"]),
+  initialize: makeAction("initialize", ["players", "setup"]),
```

Comment:
While it does work, passing a function to the initialize action will likely make our future lives difficult. We want to treat this log as an immutable record of the game's progress. On top of that immutability, we will certainly want to be able to send serialized representations of log entries across the network so other participants can append actions to their logs. So far we have preserved this property of serialization by using "flat" data structures. How might we serialize functions, though? By default they are not represented in JSON:

```es6
> JSON.stringify({huhu: 42, f: x => x + 1})
'{"huhu":42}'
```

I propose we try to keep functions (and other "rich" data structures like class instances) out of these actions. Does that make sense?